

### О задании

Практическое задание 1 посвящено изучению основных библиотек для анализа данных, а также линейных моделей и методов их обучения. Вы научитесь:
 * применять библиотеки NumPy и Pandas для осуществления желаемых преобразований;
 * подготавливать данные для обучения линейных моделей;
 * обучать линейную, Lasso и Ridge-регрессии при помощи модуля scikit-learn;
 * реализовывать обычный и стохастический градиентные спуски;
 * обучать линейную регрессию для произвольного функционала качества.
 


## Библиотеки для анализа данных

### NumPy

Во всех заданиях данного раздела запрещено использовать циклы  и list comprehensions. Под вектором и матрицей в данных заданиях понимается одномерный и двумерный numpy.array соответственно.

In [1]:
import numpy as np

**1.** Реализуйте функцию, возвращающую максимальный элемент в векторе x среди элементов, перед которыми стоит нулевой. Для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответом является 5. Если нулевых элементов нет, функция должна возвращать None.


In [19]:
# Для обработки исключений используется конструкция try - except.

def max_element_(a):
    try:
        return max(a[1:] [(a==0) [:-1]])
    except:
        return None
    
a = np.array([6, 2, 0, 33, 0, 0, 15, 7, 0]) 

print(max_element_(a))
print(a)

33
[ 6  2  0 33  0  0 15  7  0]


**2.** Реализуйте функцию, принимающую на вход матрицу и некоторое число и возвращающую ближайший к числу элемент матрицы. Например: для X = np.arange(0,10).reshape((2, 5)) и v = 3.6 ответом будет 4.

In [18]:
def nearest_value(X, v):
    return min((abs(X-v).min()) + v, X.max())
    
X = np.arange(0,10).reshape((5,2))
v = 3.6 

print(f'ближайший к числу {v} элемент матрицы -', nearest_value(X, v))
print(X)

ближайший к числу 3.6 элемент матрицы - 4.0
[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


**3.** Реализуйте функцию scale(X), которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает выборочное среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на случайной матрице (для её генерации можно использовать, например, функцию [numpy.random.randint](http://docs.scipy.org/doc/numpy/reference/generated/numpy.random.randint.html)).

In [35]:
def scale(X):
    std_deviation = X.std(axis=0)
    std_deviation[std_deviation == 0] = 1
    X = X - X.mean(axis=0) # или X -= X.mean(axis=0) c преобразованием X = X.astype(np.float)
    X /= std_deviation
    return(X)

X = np.random.randint(1, 10, (2,5))
print('Матрица:', X, 'Масштабированная матрица:', scale(X), sep='\n')

Матрица:
[[7 7 6 8 3]
 [8 3 6 2 3]]
Масштабированная матрица:
[[-1.  1.  0.  1.  0.]
 [ 1. -1.  0. -1.  0.]]


**4.** Реализуйте функцию, которая для заданной матрицы находит:
 - определитель
 - след
 - наименьший и наибольший элементы
 - норму Фробениуса
 - собственные числа
 - обратную матрицу

Для тестирования сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}$(10,1)

In [44]:
def get_stats(X_4):
    print('Определитель матрицы :','\n', np.linalg.det(X_4), '\n', # https://pyprog.pro/linear_algebra_functions/linalg_det.html
          'След матрицы :','\n', X_4.trace(), '\n',  # https://pyprog.pro/linear_algebra_functions/trace.html
          'Наименьший элемент :','\n', X_4.min(), '\n',
          'Наибольший элемент :','\n', X_4.max(), '\n',
          'Норма Фробениуса :','\n', np.linalg.norm(X_4), '\n', # https://pyprog.pro/linear_algebra_functions/linalg_norm.html
          'Собственные числа :','\n', np.linalg.eigvals(X_4),'\n', # https://pyprog.pro/linear_algebra_functions/linalg_eigvals.html
          'Обратная матрица :','\n', np.linalg.inv(X_4), '\n') # https://pyprog.pro/linear_algebra_functions/linalg_inv.html
        
X_4 = np.random.normal(10, 1, (2, 2))  # https://pyprog.pro/linear_algebra_functions/linalg_norm.html

print(X_4)
print(get_stats(X_4))

[[10.47081606 11.86210301]
 [10.35923284 12.54294653]]
Определитель матрицы : 
 8.452598894445531 
 След матрицы : 
 23.013762588842127 
 Наименьший элемент : 
 10.35923283668692 
 Наибольший элемент : 
 12.542946527753342 
 Норма Фробениуса : 
 22.693318168098934 
 Собственные числа : 
 [ 0.37334106 22.64042153] 
 Обратная матрица : 
 [[ 1.48391597 -1.40336755]
 [-1.22556778  1.23876883]] 

None


**5.** Повторите 100 раз следующий эксперимент: сгенерируйте две матрицы размера 10×10 из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

In [ ]:
for exp_num in range(100):
    # Your code here

### Pandas

![](https://metrouk2.files.wordpress.com/2015/10/panda.jpg)

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

[Данные](https://www.dropbox.com/s/dvfitn93obn0rql/2008.csv?dl=0) и их [описание](http://stat-computing.org/dataexpo/2009/the-data.html)

In [2]:
import pandas as pd
%matplotlib inline

In [3]:
air_flights = pd.read_csv('2008.csv') #загружка данных

#head()-первые строки, tail()-последние строки, sample()-случайные строки !!! по умолчанию выводит 5 строк
air_flights.head(10) 

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      6          18          3   2111.0        2055   2308.0   
1  2008      6           4          3   1426.0        1410   1730.0   
2  2008      8           3          7   1143.0        1145   1501.0   
3  2008      1          23          3   1141.0        1144   1323.0   
4  2008      5           4          7    815.0         820   1243.0   
5  2008      1           3          4   1349.0        1325   1640.0   
6  2008      3          31          1   1359.0        1400   1413.0   
7  2008      4          22          2   1815.0        1820   2057.0   
8  2008      3           4          2    626.0         630    820.0   
9  2008     10           6          1   1905.0        1907     24.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2300            WN        242  ...    3.0      8.0          0   
1        1728            XE       2380  ...    4.0     12.0          0   
2        1520            WN       1769  ...    4.0      9.0          0   
3        1335            OO       3802  ...    4.0     19.0          0   
4        1300            WN        399  ...    4.0      8.0          0   
5        1609            B6        834  ...    5.0     13.0          0   
6        1415            WN       1589  ...    5.0     14.0          0   
7        2131            DL        617  ...    9.0     31.0          0   
8         830            WN        454  ...    4.0      6.0          0   
9          37            UA        270  ...    6.0     16.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0           NaN          NaN      NaN   
1               NaN         0           NaN          NaN      NaN   
2               NaN         0           NaN          NaN      NaN   
3               NaN         0           NaN          NaN      NaN   
4               NaN         0           NaN          NaN      NaN   
5               NaN         0          13.0          0.0      7.0   
6               NaN         0           NaN          NaN      NaN   
7               NaN         0           NaN          NaN      NaN   
8               NaN         0           NaN          NaN      NaN   
9               NaN         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4            NaN                NaN  
5            0.0               11.0  
6            NaN                NaN  
7            NaN                NaN  
8            NaN                NaN  
9            NaN                NaN  

[10 rows x 29 columns]

In [61]:
#info()-общая информация по датафрейму, тип, так же видно пропуски в данных
air_flights.info()  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Year               70000 non-null  int64  
 1   Month              70000 non-null  int64  
 2   DayofMonth         70000 non-null  int64  
 3   DayOfWeek          70000 non-null  int64  
 4   DepTime            68601 non-null  float64
 5   CRSDepTime         70000 non-null  int64  
 6   ArrTime            68444 non-null  float64
 7   CRSArrTime         70000 non-null  int64  
 8   UniqueCarrier      70000 non-null  object 
 9   FlightNum          70000 non-null  int64  
 10  TailNum            69180 non-null  object 
 11  ActualElapsedTime  68418 non-null  float64
 12  CRSElapsedTime     69991 non-null  float64
 13  AirTime            68418 non-null  float64
 14  ArrDelay           68418 non-null  float64
 15  DepDelay           68601 non-null  float64
 16  Origin             700

*	1	*	Year	****	1987-2008	****	1987-2008
*	2	*	Month	****	1-12	****	
*	3	*	DayofMonth	****	1-31	****	
*	4	*	DayOfWeek	****	1 (Monday) - 7 (Sunday)	****	1 (понедельник) - 7 (воскресенье)
*	5	*	DepTime	****	actual departure time (local, hhmm)	****	фактическое время отправления (местное, hhmm)
*	6	*	CRSDepTime	****	scheduled departure time (local, hhmm)	****	запланированное время отправления (местное, hhmm)
*	7	*	ArrTime	****	actual arrival time (local, hhmm)	****	фактическое время прибытия (местное, hhmm)
*	8	*	CRSArrTime	****	scheduled arrival time (local, hhmm)	****	запланированное время прибытия (местное, hhmm)
*	9	*	UniqueCarrier	****	unique carrier code	****	уникальный код перевозчика
*	10	*	FlightNum	****	flight number	****	номер рейса
*	11	*	TailNum	****	plane tail number	****	хвостовой номер самолета
*	12	*	ActualElapsedTime	****	in minutes	****	в считанные минуты
*	13	*	CRSElapsedTime	****	in minutes	****	в считанные минуты
*	14	*	AirTime	****	in minutes	****	в считанные минуты
*	15	*	ArrDelay	****	arrival delay, in minutes	****	задержка прибытия, в минутах
*	16	*	DepDelay	****	departure delay, in minutes	****	задержка отправления, в минутах
*	17	*	Origin	****	origin IATA airport code	****	происхождение кода аэропорта IATA
*	18	*	Dest	****	destination IATA airport code	****	код аэропорта назначения IATA
*	19	*	Distance	****	in miles	****	в милях
*	20	*	TaxiIn	****	taxi in time, in minutes	****	такси вовремя, в считанные минуты
*	21	*	TaxiOut	****	taxi out time in minutes	****	время выезда на такси в минутах
*	22	*	Cancelled	****	was the flight cancelled?	****	был ли отменен рейс?
*	23	*	CancellationCode	****	reason for cancellation (A = carrier, B = weather, C = NAS, D = security)	****	причина отмены (A = перевозчик, B = погода, C = NAS, D = безопасность)
*	24	*	Diverted	****	1 = yes, 0 = no	****	1 = да, 0 = нет
*	25	*	CarrierDelay	****	in minutes	****	в считанные минуты
*	26	*	WeatherDelay	****	in minutes	****	в считанные минуты
*	27	*	NASDelay	****	in minutes	****	в считанные минуты
*	28	*	SecurityDelay	****	in minutes	****	в считанные минуты
*	29	*	LateAircraftDelay	****	in minutes	****	в считанные минуты


In [12]:
air_flights.describe(include = 'all') # статистика включает не только числовые столбцы, но и строки 
# (unique - сколько уникальных значений, top - какое самое популярное значение, freq - как часто встречается самое популярное)

Year         Month    DayofMonth     DayOfWeek       DepTime  \
count   70000.0  70000.000000  70000.000000  70000.000000  68601.000000   
unique      NaN           NaN           NaN           NaN           NaN   
top         NaN           NaN           NaN           NaN           NaN   
freq        NaN           NaN           NaN           NaN           NaN   
mean     2008.0      6.380300     15.742300      3.925900   1331.025058   
std         0.0      3.419011      8.813738      1.995955    477.005973   
min      2008.0      1.000000      1.000000      1.000000      1.000000   
25%      2008.0      3.000000      8.000000      2.000000    928.000000   
50%      2008.0      6.000000     16.000000      4.000000   1323.000000   
75%      2008.0      9.000000     23.000000      6.000000   1725.000000   
max      2008.0     12.000000     31.000000      7.000000   2400.000000   

          CRSDepTime       ArrTime    CRSArrTime UniqueCarrier     FlightNum  \
count   70000.000000  68444.000000  70000.000000         70000  70000.000000   
unique           NaN           NaN           NaN            20           NaN   
top              NaN           NaN           NaN            WN           NaN   
freq             NaN           NaN           NaN         12028           NaN   
mean     1324.217071   1478.600842   1492.904771           NaN   2214.752114   
std       463.231562    504.601176    482.071631           NaN   1963.606525   
min         1.000000      1.000000      1.000000           NaN      1.000000   
25%       926.000000   1106.000000   1115.000000           NaN    614.000000   
50%      1316.000000   1510.000000   1515.000000           NaN   1562.000000   
75%      1715.000000   1906.000000   1905.000000           NaN   3500.000000   
max      2359.000000   2400.000000   2400.000000           NaN   7828.000000   

        ...        TaxiIn       TaxiOut     Cancelled  CancellationCode  \
count   ...  68444.000000  68594.000000  70000.000000              1411   
unique  ...           NaN           NaN           NaN                 3   
top     ...           NaN           NaN           NaN                 A   
freq    ...           NaN           NaN           NaN               563   
mean    ...      6.847189     16.435621      0.020157               NaN   
std     ...      4.884862     11.297027      0.140539               NaN   
min     ...      0.000000      0.000000      0.000000               NaN   
25%     ...      4.000000     10.000000      0.000000               NaN   
50%     ...      6.000000     14.000000      0.000000               NaN   
75%     ...      8.000000     19.000000      0.000000               NaN   
max     ...    165.000000    317.000000      1.000000               NaN   

            Diverted  CarrierDelay  WeatherDelay      NASDelay  SecurityDelay  \
count   70000.000000  15253.000000  15253.000000  15253.000000   15253.000000   
unique           NaN           NaN           NaN           NaN            NaN   
top              NaN           NaN           NaN           NaN            NaN   
freq             NaN           NaN           NaN           NaN            NaN   
mean        0.002443     15.862388      2.829935     16.751918       0.078935   
std         0.049365     38.596304     17.687256     32.017588       2.063460   
min         0.000000      0.000000      0.000000      0.000000       0.000000   
25%         0.000000      0.000000      0.000000      0.000000       0.000000   
50%         0.000000      0.000000      0.000000      5.000000       0.000000   
75%         0.000000     17.000000      0.000000     21.000000       0.000000   
max         1.000000   1146.000000    516.000000    914.000000     175.000000   

        LateAircraftDelay  
count        15253.000000  
unique                NaN  
top                   NaN  
freq                  NaN  
mean            20.698944  
std             39.068022  
min              0.000000  
25%              0.000000  
50%              0.0000

**6.** Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)

In [58]:
# value_counts()- вернет количество уникальных совпадений в определенной колонке,в порядке убывания.
air_flights.CancellationCode.value_counts() 

A    563
B    549
C    299
Name: CancellationCode, dtype: int64

причина отмены (A- 563)--> A = перевозчик

**7.** Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.

In [10]:
# предположим эти данные находятся в "Distance" (в милях)
print('Среднее пройденное расстояние- ', air_flights.Distance.mean())
print('Минимальное пройденное расстояние- ', air_flights.Distance.min())
print('Максимальное пройденное расстояние- ', air_flights.Distance.max())

Среднее пройденное расстояние-  724.5082571428571
Минимальное пройденное расстояние-  31
Максимальное пройденное расстояние-  4962


**8.** Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?

В какие дни и на каких рейсах оно было?
* на примере первой строки- 2008 (год), 12 месяц (декабрь), 30 (число), 2 (вторник), 65 (№ рейса), 31 (пройденное расстояние)

In [13]:
# т.к. уже известно зн. (31) по нему выведем необходимые нам данные 
air_flights[air_flights.Distance == 31][['Year','Month','DayofMonth', 'DayOfWeek', 'FlightNum', 'Distance']]

Year  Month  DayofMonth  DayOfWeek  FlightNum  Distance
1116   2008     12          30          2         65        31
6958   2008     12          26          5         65        31
17349  2008      8          18          1         64        31
27534  2008      3          11          2         64        31
46082  2008      8           9          6         65        31
48112  2008      2          28          4         64        31

Какое расстояние было пройдено этими же рейсами в другие дни?
* нас интерессуют рейсы №64 и №64 (FlightNum - номер рейса)

In [49]:
pd.unique(air_flights[air_flights.FlightNum.isin([64, 65])]['Distance'])

array([ 533,   31,  680, 2381,   82, 1005,  123,  359, 1747,  571,  372,
       2454,  581,  883,  281,  682,  414,  386], dtype=int64)

 Какое расстояние было пройдено этими же рейсами в другие дни?
 * 

**9.** Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?

In [16]:
air_flights['Origin'].value_counts()

ATL    4134
ORD    3550
DFW    2793
DEN    2383
LAX    2082
       ... 
AKN       1
PLN       1
LWB       1
BLI       1
ADK       1
Name: Origin, Length: 297, dtype: int64

больше всего вылетов (4134) из аэропорта - ATL (Город: Atlanta)

**10.** Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?

* для каждого аэропорта среднее время полета-сгруппируем по столбцу Origin, значение будет AirTime -среднее

In [47]:
mean_time = air_flights.groupby('Origin')[['AirTime']].mean()
mean_time

AirTime
Origin           
ABE     88.266667
ABI     36.400000
ABQ     93.454321
ABY     35.714286
ACK     50.800000
...           ...
WRG     18.000000
XNA     85.945736
YAK     35.900000
YKM     79.000000
YUM     47.470588

[297 rows x 1 columns]

* получилось 297 записей со средними значениями полета каждого аэропорта, отсортируем по убыванию 
* (ascending = False - по убыванию, True-по возрастанию)

In [53]:
mean_time.sort_values(by = ['AirTime'], ascending = False)

AirTime
Origin            
SJU     205.200000
BQN     193.954545
PSE     191.875000
STT     185.029412
ANC     165.415094
...            ...
PSG      20.250000
OXR      19.666667
WRG      18.000000
GST      13.000000
DLG            NaN

[297 rows x 1 columns]

Какой аэропорт имеет наибольшее значение этого показателя?
* как видим наибольшее среднее время полета было из аэропорта:
* SJU = 205.200000 (Страна: Пуэрто Рико (Puerto Rico), Город: Сан Хуан (San Juan))

**11.** Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [127]:

w = air_flights[(air_flights['DepDelay'] > 0) & (air_flights['FlightNum'] > 1000)]
#w_1 = w['FlightNum'].sort_values(ascending = False)
w['Origin'].value_counts()

ORD    1065
ATL    1004
DFW     987
IAH     533
LAX     478
       ... 
ACY       1
RFD       1
TEX       1
FLO       1
STX       1
Name: Origin, Length: 260, dtype: int64

## Линейная регрессия

В этой части мы разберемся с линейной регрессией, способами её обучения и измерением качества ее прогнозов. 

Будем рассматривать датасет из предыдущей части задания для предсказания времени задержки отправления рейса в минутах (DepDelay). Отметим, что под задержкой подразумевается не только опоздание рейса относительно планируемого времени вылета, но и отправление до планируемого времени.

### Подготовка данных

**12.** Считайте выборку из файла при помощи функции pd.read_csv и ответьте на следующие вопросы:
   - Имеются ли в данных пропущенные значения?
   - Сколько всего пропущенных элементов в таблице "объект-признак"?
   - Сколько объектов имеют хотя бы один пропуск?
   - Сколько признаков имеют хотя бы одно пропущенное значение?

In [35]:
air_flights.info()  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Year               70000 non-null  int64  
 1   Month              70000 non-null  int64  
 2   DayofMonth         70000 non-null  int64  
 3   DayOfWeek          70000 non-null  int64  
 4   DepTime            68601 non-null  float64
 5   CRSDepTime         70000 non-null  int64  
 6   ArrTime            68444 non-null  float64
 7   CRSArrTime         70000 non-null  int64  
 8   UniqueCarrier      70000 non-null  object 
 9   FlightNum          70000 non-null  int64  
 10  TailNum            69180 non-null  object 
 11  ActualElapsedTime  68418 non-null  float64
 12  CRSElapsedTime     69991 non-null  float64
 13  AirTime            68418 non-null  float64
 14  ArrDelay           68418 non-null  float64
 15  DepDelay           68601 non-null  float64
 16  Origin             700

*	1	*	Year	****	1987-2008	****	1987-2008
*	2	*	Month	****	1-12	****	
*	3	*	DayofMonth	****	1-31	****	
*	4	*	DayOfWeek	****	1 (Monday) - 7 (Sunday)	****	1 (понедельник) - 7 (воскресенье)
*	5	*	DepTime	****	actual departure time (local, hhmm)	****	фактическое время отправления (местное, hhmm)
*	6	*	CRSDepTime	****	scheduled departure time (local, hhmm)	****	запланированное время отправления (местное, hhmm)
*	7	*	ArrTime	****	actual arrival time (local, hhmm)	****	фактическое время прибытия (местное, hhmm)
*	8	*	CRSArrTime	****	scheduled arrival time (local, hhmm)	****	запланированное время прибытия (местное, hhmm)
*	9	*	UniqueCarrier	****	unique carrier code	****	уникальный код перевозчика
*	10	*	FlightNum	****	flight number	****	номер рейса
*	11	*	TailNum	****	plane tail number	****	хвостовой номер самолета
*	12	*	ActualElapsedTime	****	in minutes	****	в считанные минуты
*	13	*	CRSElapsedTime	****	in minutes	****	в считанные минуты
*	14	*	AirTime	****	in minutes	****	в считанные минуты
*	15	*	ArrDelay	****	arrival delay, in minutes	****	задержка прибытия, в минутах
*	16	*	DepDelay	****	departure delay, in minutes	****	задержка отправления, в минутах
*	17	*	Origin	****	origin IATA airport code	****	происхождение кода аэропорта IATA
*	18	*	Dest	****	destination IATA airport code	****	код аэропорта назначения IATA
*	19	*	Distance	****	in miles	****	в милях
*	20	*	TaxiIn	****	taxi in time, in minutes	****	такси вовремя, в считанные минуты
*	21	*	TaxiOut	****	taxi out time in minutes	****	время выезда на такси в минутах
*	22	*	Cancelled	****	was the flight cancelled?	****	был ли отменен рейс?
*	23	*	CancellationCode	****	reason for cancellation (A = carrier, B = weather, C = NAS, D = security)	****	причина отмены (A = перевозчик, B = погода, C = NAS, D = безопасность)
*	24	*	Diverted	****	1 = yes, 0 = no	****	1 = да, 0 = нет
*	25	*	CarrierDelay	****	in minutes	****	в считанные минуты
*	26	*	WeatherDelay	****	in minutes	****	в считанные минуты
*	27	*	NASDelay	****	in minutes	****	в считанные минуты
*	28	*	SecurityDelay	****	in minutes	****	в считанные минуты
*	29	*	LateAircraftDelay	****	in minutes	****	в считанные минуты


- Имеются ли в данных пропущенные значения?

In [43]:
# isna() для проверки наличия пропущенных значений
air_flights.isna().sum()

Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime               1399
CRSDepTime               0
ArrTime               1556
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                820
ActualElapsedTime     1582
CRSElapsedTime           9
AirTime               1582
ArrDelay              1582
DepDelay              1399
Origin                   0
Dest                     0
Distance                 0
TaxiIn                1556
TaxiOut               1406
Cancelled                0
CancellationCode     68589
Diverted                 0
CarrierDelay         54747
WeatherDelay         54747
NASDelay             54747
SecurityDelay        54747
LateAircraftDelay    54747
dtype: int64

- Сколько всего пропущенных элементов в таблице "объект-признак"?

In [41]:
air_flights.isna().sum().sum()

355215

- Сколько объектов имеют хотя бы один пропуск?

In [57]:
# метод к строке (axis=0) или столбцу (axis=1)
(air_flights.isna().sum(axis=1) > 0).sum()

70000

- Сколько признаков имеют хотя бы одно пропущенное значение?

In [58]:
(air_flights.isna().sum(axis=0) > 0).sum()

16

Как вы понимаете, также не имеет смысла рассматривать при решении поставленной задачи объекты с пропущенным значением целевой переменной. В связи с этим ответьте на следующие вопросы и выполните соответствующие действия:
- Имеются ли пропущенные значения в целевой переменной?
- Проанализируйте объекты с пропущенными значениями целевой переменной. Чем вызвано это явление? Что их объединяет? Можно ли в связи с этим, на ваш взгляд, исключить какие-то признаки из рассмотрения? Обоснуйте свою точку зрения.

Исключите из выборки объекты **с пропущенным значением целевой переменной и со значением целевой переменной, равным 0**, а также при необходимости исключите признаки в соответствии с вашим ответом на последний вопрос из списка и выделите целевую переменную в отдельный вектор, исключив её из матрицы "объект-признак".

In [ ]:
# Your code here

**13.** Обратите внимание, что признаки DepTime, CRSDepTime, ArrTime, CRSArrTime приведены в формате hhmm, в связи с чем будет не вполне корректно рассматривать их как вещественные.

Преобразуйте каждый признак FeatureName из указанных в пару новых признаков FeatureName\_Hour, FeatureName\_Minute, разделив каждое из значений на часы и минуты. Не забудьте при этом исключить исходный признак из выборки. В случае, если значение признака отсутствует, значения двух новых признаков, его заменяющих, также должны отсутствовать. 

Например, признак DepTime необходимо заменить на пару признаков DepTime_Hour, DepTime_Minute. При этом, например, значение 155 исходного признака будет преобразовано в значения 1 и 55 признаков DepTime_Hour, DepTime_Minute соответственно.

In [ ]:
# Your code here

**14.** Некоторые из признаков, отличных от целевой переменной, могут оказывать чересчур значимое влияние на прогноз, поскольку по своему смыслу содержат большую долю информации о значении целевой переменной. Изучите описание датасета и исключите признаки, сильно коррелирующие с ответами. Ваш выбор признаков для исключения из выборки обоснуйте. Кроме того, исключите признаки TailNum и Year.

In [ ]:
# Your code here

**15.** Приведем данные к виду, пригодному для обучения линейных моделей. Для этого вещественные признаки надо отмасштабировать, а категориальные — привести к числовому виду. Также надо устранить пропуски в данных.

В первую очередь поймем, зачем необходимо применять масштабирование. Следующие ячейки с кодом построят гистограммы для 3 вещественных признаков выборки.

In [ ]:
X['DepTime_Hour'].hist(bins=20)

In [ ]:
X['TaxiIn'].hist(bins=20)

In [ ]:
X['FlightNum'].hist(bins=20)

Какую проблему вы наблюдаете на этих графиках? Как масштабирование поможет её исправить?

Некоторые из признаков в нашем датасете являются категориальными. Типичным подходом к работе с ними является бинарное, или [one-hot-кодирование](https://en.wikipedia.org/wiki/One-hot).

Реализуйте функцию transform_data, которая принимает на вход DataFrame с признаками и выполняет следующие шаги:
1. Замена пропущенных значений на нули для вещественных признаков и на строки 'nan' для категориальных.
2. Масштабирование вещественных признаков с помощью [StandardScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html).
3. One-hot-кодирование категориальных признаков с помощью [DictVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html) или функции [pd.get_dummies](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html).

Метод должен возвращать преобразованный DataFrame, который должна состоять из масштабированных вещественных признаков и закодированных категориальных (исходные признаки должны быть исключены из выборки).

In [ ]:
def transform_data(data):
    # Your code here

Примените функцию transform_data к данным. Сколько признаков получилось после преобразования?

In [ ]:
# Your code here

**16. (0.5 балла)** Разбейте выборку и вектор целевой переменной на обучение и контроль в отношении 70/30 (для этого можно использовать, например, функцию [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.train_test_split.html)). 

In [ ]:
# Your code here

### Scikit-learn

<img src = "https://pp.vk.me/c4534/u35727827/93547647/x_d31c4463.jpg">
Теперь, когда мы привели данные к пригодному виду, попробуем решить задачу при помощи метода наименьших квадратов. Напомним, что данный метод заключается в оптимизации функционала $MSE$:

$$MSE(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 \to \min_{w},$$

где $\{ (x_i, y_i ) \}_{i=1}^l$ — обучающая выборка, состоящая из $l$ пар объект-ответ.

Заметим, что решение данной задачи уже реализовано в модуле sklearn в виде класса [LinearRegression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression).

**17.** Обучите линейную регрессию на 1000 объектах из обучающей выборки и выведите значения $MSE$ и $R^2$ на этой подвыборке и контрольной выборке (итого 4 различных числа). Проинтерпретируйте полученный результат — насколько качественные прогнозы строит полученная модель? Какие проблемы наблюдаются в модели?

**Подсказка**: изучите значения полученных коэффициентов $w$, сохраненных в атрибуте coef_ объекта LinearRegression.

In [ ]:
# Your code here

Для решения описанных вами в предыдущем пункте проблем используем L1- или L2-регуляризацию, тем самым получив Lasso и Ridge регрессии соответственно и изменив оптимизационную задачу одним из следующих образов:
$$MSE_{L1}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_1 \to \min_{w},$$
$$MSE_{L2}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_2^2 \to \min_{w},$$

где $\alpha$ — коэффициент регуляризации. Один из способов его подбора заключается в переборе некоторого количества значений и оценке качества на кросс-валидации для каждого из них, после чего выбирается значение, для которого было получено наилучшее качество.

**18.** Обучите линейные регрессии с L1- и L2-регуляризатором, подобрав лучшее значение параметра регуляризации из списка alpha_grid при помощи кросс-валидации c 5 фолдами на тех же 1000 объектах, что и в п.17. Выведите значения $MSE$ и $R^2$ на обучающей и контрольной выборках. Удалось ли решить указанные вами ранее проблемы?

Для выполнения данного задания вам могут понадобиться реализованные в библиотеке объекты [LassoCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html), [RidgeCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html) и [KFold](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.KFold.html).


In [ ]:
# Your code here

### Градиентный спуск

В предыдущем разделе мы использовали существующие реализации методов обучения линейной регрессии с регуляризацией и без. Тем не менее, подобные реализации, как правило, имеются лишь для ограниченного набора стандартных методов. В частности, при выходе функционала качества за пределы стандартного множества необходимо самостоятельно реализовывать составляющие процесса решения оптимизационной задачи. Именно этому и посвящен данный раздел задания.

Пусть необходимо минимизировать следующий функционал (Mean Square Percentage Error — модифицированный [RMSPE](https://www.kaggle.com/c/rossmann-store-sales/details/evaluation)):
$$MSPE(\{x_i, y_i\}_{i=1}^l, \, w) = \frac{1}{l}\sum_{i=1}^l \left( \frac{y_i - \langle w, x_i \rangle }{y_i} \right)^2,$$

где $\{x_i, y_i\}_{i=1}^l$ — обучающая выборка, $w$ — вектор весов линейной модели. Будем также рассматривать функционал $MSPE$ с L2-регуляризацией:

$$MSPE(\{x_i, y_i\}_{i=1}^l, \, w) = \frac{1}{l}\sum_{i=1}^l \left( \frac{y_i - \langle w, x_i \rangle }{y_i} \right)^2 + ||w||_2^2.$$

**19.** Добавьте к объектам обеих выборок из п. 16 единичный признак.

In [ ]:
# Your code here

**20.** Реализуйте функции, которые вычисляют:
 * прогнозы линейной модели;
 * функционал $MSPE$ и его градиент;
 * регуляризованный $MSPE$ и его градиент.

In [ ]:
# возвращает вектор прогнозов линейной модели с вектором весов w для выборки X
def make_pred(X, w):
    pass

In [ ]:
# возвращает значение функционала MSPE для выборки (X, y) и вектора весов w
def get_func(w, X, y):
    pass

In [ ]:
# возвращает градиент функционала MSPE для выборки (X, y) и вектора весов w
def get_grad(w, X, y):
    pass

In [ ]:
# возвращает значение регуляризованного функционала MSPE для выборки (X, y) и вектора весов w
def get_reg_func(w, X, y):
    pass

In [ ]:
# возвращает градиент регуляризованного функционала MSPE для выборки (X, y) и вектора весов w
def get_reg_grad(w, X, y):
    pass

**21.** Реализуйте метод градиентного спуска для описанных функционалов ($MSPE$ и его регуляризованный вариант). Функция должна принимать следующие параметры:
 - X — матрица "объект-признак";
 - y — вектор целевой переменной;
 - w0 — начальное значение вектора весов;
 - step_size — значение темпа обучения;
 - max_iter — максимальное число итераций;
 - eps — значение, используемое в критерии останова;
 - is_reg — бинарный параметр, принимает значение True в случае наличия регуляризации функционала, False — в противном случае.
 
Процесс должен быть остановлен, если выполнено хотя бы одно из следующих условий:
 - было выполнено заданное количество итераций max_iter;
 - евклидова норма разности векторов $w$ на соседних итерациях стала меньше, чем eps.

Функция должна возвращать полученный в результате оптимизации вектор $w$ и список значений функционала на каждой итерации.

In [ ]:
def grad_descent(X, y, step_size, max_iter, eps, is_reg):
    # Your code here

Обучите линейную регрессию с функционалом $MSPE$ на обучающей выборке при помощи метода градиентного спуска и изобразите кривые зависимости значения функционала от номера итерации для различных:
 * значений размера шага из набора [0.001, 1, 10];
 * способов начальной инициализации вектора весов (нули, случайные веса).

Проанализируйте полученные результаты — влияют ли данные параметры на скорость сходимости и итоговое качество? Если да, то как?

In [ ]:
# Your code here

**22.** Обучите линейную регрессию с функционалом MSPE и его регуляризованным вариантом на обучающей выборке при помощи метода градиентного спуска и изобразите кривые зависимости значения функционала от номера итерации. Исследуйте зависимость скорости сходимости от наличия регуляризации. Обоснуйте, почему так происходит.

In [ ]:
# Your code here

Метод градиентного спуска может быть весьма трудозатратен в случае большого размера обучающей выборки. Поэтому часто используют метод стохастического градиентного спуска, где на каждой итерации выбирается случайный объект из обучающей выборки и обновление весов происходит только по этому объекту. 

**23.**  Реализуйте метод стохастического градиентного спуска (SGD) для описанных функционалов ($MSPE$ и его регуляризованный вариант). Функция должна иметь параметры и возвращаемое значение, аналогичные оным функции grad\_descent из п.21. Кроме того, должен использоваться аналогичный критерий останова.

In [ ]:
def sgd(X, y, step_size, max_iter, eps, is_reg):
    # Your code here

Обучите линейную регрессию с функционалом $MSPE$ и его регуляризованным вариантом на обучающей выборке при помощи метода стохастического градиентного спуска, подобрав при этом размер шага, при котором метод будет сходиться. Нарисуйте график сходимости. Выведите значения $MSPE, MSE, R^2$ на контрольной выборке.

In [ ]:
# Your code here

**24.** Аналогично п.22 исследуйте зависимость скорости сходимости метода SGD от наличия регуляризации. Обоснуйте, почему так происходит.

In [ ]:
# Your code here

**25.** Обучите стандартную линейную регрессию с функционалом качества MSE на обучающей выборке и выведите значение MSPE полученного решения на контрольной выборке. Как оно соотносится с аналогичным результатом для решения, полученного в п.22? Почему?

In [ ]:
# Your code here